modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/IgniteSqlValidator.java

```diff
@@ -557,82 +585,93 @@ private boolean isSystemFieldName(String alias) {
+        RelDataType parameterType;
+        // IgniteCustomType: first we must check whether dynamic parameter is a custom data type.
+        // If so call createCustomType with appropriate arguments.
+        if (value instanceof UUID) {
+            parameterType = typeFactory.createCustomType(UuidType.NAME);
+        } else if (value == null) {
+            parameterType = typeFactory.createSqlType(SqlTypeName.NULL);
+        } else {
+            parameterType = typeFactory.toSql(typeFactory.createType(value.getClass()));
+        }
 
-            Arrays.fill(operandTypes, unknownType);
+        // Dynamic parameters are always nullable.
+        // Otherwise it seem to cause "Conversion to relational algebra failed to preserve datatypes" errors
+        // in some cases.
+        return typeFactory.createTypeWithNullability(parameterType, true);
+    }
 
-            if (operandTypeInference != null && !(call instanceof SqlCase)) {
-                operandTypeInference.inferOperandTypes(callBinding, inferredType, operandTypes);
-            } else if (operandTypeChecker instanceof FamilyOperandTypeChecker) {
-                // Infer operand types from checker for dynamic parameters if it's possible.
-                FamilyOperandTypeChecker checker = (FamilyOperandTypeChecker) operandTypeChecker;
+    private Object getDynamicParamValue(SqlDynamicParam dynamicParam) {
+        Object value = parameters[dynamicParam.getIndex()];
+        // save dynamic parameter for later validation.
+        dynamicParams[dynamicParam.getIndex()] = dynamicParam;
+        return value;
+    }
 
-                for (int i = 0; i < checker.getOperandCountRange().getMax(); i++) {
-                    if (i >= operandTypes.length) {
-                        break;
-                    }
+    /**
+     * Post validation check of dynamic parameters.
+     */
+    private void checkDynamicParameters() {
+        // Derived types of dynamic parameters should not change (current type inference behavior).
 
-                    SqlTypeFamily family = checker.getOperandSqlTypeFamily(i);
-                    RelDataType type = family.getDefaultConcreteType(typeFactory());
+        for (int i = 0; i < parameters.length; i++) {
+            SqlDynamicParam param = dynamicParams[i];
+            assert param != null : format("Dynamic parameter#{} has not been checked", i);
 
-                    if (type != null && operands.get(i) instanceof SqlDynamicParam) {
-                        operandTypes[i] = type;
-                    }
-                }
-            }
+            RelDataType paramType = getDynamicParamType(param);
+            RelDataType derivedType = getValidatedNodeType(param);
 
-            for (int i = 0; i < operands.size(); ++i) {
-                SqlNode operand = operands.get(i);
+            // We can check for nullability, but it was set to true.
+            if (!SqlTypeUtil.equalSansNullability(derivedType, paramType)) {
```

Comment:
you need to use org.apache.ignite.internal.util.IgniteUtils#assertionsEnabled or rework this approach at all